viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py

```diff
@@ -0,0 +1,85 @@
+import atexit
+import os
+import signal
+import socket
+import subprocess
+import sys
+import threading
+import time
+
+import server
+
+from typing import Optional
+
+
+def is_port_open(port: int):
+    # check whether the given port is open
+    try:
+        sock = socket.socket()
+        sock.bind(("", port))
+        sock.close()
+
+        return True
+    except OSError:
+        return False
+
+
+def get_free_port(default_port: int = None):
+    if default_port:
+        if is_port_open(default_port):
+            return default_port
+    sock = socket.socket()
+    sock.bind(("", 0))
+    port = sock.getsockname()[1]
+
+    return port
+
+
+def run_bridge_server_as_subprocess(
+    websocket_port: int,
+    zmq_port: Optional[int] = None,
```

Comment:
ChatGPT:
In Python, Union[int, None] and Optional[int] are similar in that they both indicate that a variable can be either an integer or None. However, there is a subtle difference between them.

Union[int, None] explicitly specifies that the variable can be either an integer or None. It does not allow any other type to be assigned to it. On the other hand, Optional[int] is actually an alias for Union[int, None], which means it represents the same set of possible types. However, Optional[int] does not explicitly indicate that the variable cannot be assigned any other type.

Therefore, using Union[int, None] is more specific and explicit, and may help prevent type-related bugs in your code. However, both Optional[int] and Union[int, None] can be used interchangeably in most cases.

Reply:
Since Optional[int] is equivalent to Union[int, None] whereas more concise and readable, we're using Optional as the type annotation.